### PR TITLE
Stats: Add support for CSV download of UTM data

### PR DESCRIPTION
**Pull request by a8ck3n**
Related to #87687 

## Proposed Changes

Adds the "Download data as CSV" button to the UTM Summary page.

Differs from existing download buttons as they rely on the Redux setup with it's query component. Instead we use the data as fetched via our custom hook and generate CSV data based on that.

### Screenshot

SCR-20240318-spzr

## Testing Instructions

* Launch the Live site.
* Visit the Traffic page on a site with UTM data.
* Enable the `stats/utm-module` flag.
* Scroll down to the UTM section and click "View all" or "View details".
* Or go directly to Summary page at: `/stats/day/utm/SITE_SLUG`
* Confirm the download button is enabled and works if there is UTM data to show.
* Click the download button and confirm the CSV data looks correct.
* Change the period of the request.
* Download again and confirm the CSV data reflects the updated data.
* Confirm the button is disabled if there are no results displayed.

Note: If you revert the last commit on the branch, you can test the module with Jetpack.com which has lots of UTM data to test with.

### Odd behaviour and a question

In my testing against Jetpack.com, I'm seeing odd results when requesting data via campaign. The API returns some unlabeled campaigns with children per this screenshot:

SCR-20240318-synw

I'm not sure how this is even possible but on the export side, this currently results in exports like the following:

- `"undefined > post title",COUNT_VALUE`

I looked at testing the campaign and returning an empty string but I'm not sure that is a better solution as it might not be immediately obvious that you have bad data. At least with the "undefined" string, it's easy to spot these rows and either delete or ignore as appropriate.

Any thoughts on if this should be handled differently, let me know!

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/my-sites/stats/stats-module-utm/utm-export-button.jsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	const shouldDisableExport = data && data.length !== 0 ? false : true;
+
+	// Turns the working data into a flattened array of objects.
+	// Preserves the original data but updates the label for export.
+	const flattenDataForExport = ( originalData ) => {
+		const newData = [];
+		// Map the array of objects.
+		originalData.forEach( ( row ) => {
+			// Wrap label in quotes and push values.
+			let newLabel = `"${ row.label }"`;
```
Comment (kangzj) [inline client/my-sites/stats/stats-module-utm/utm-export-button.jsx:18]: Just to be safe, we probably want to deal with any double quotes in `row.label` before quoting it, example:

https://github.com/Automattic/wp-calypso/blob/ae401a84d6e9fe28cefc9dc4fd3045f4891d1746/client/my-sites/stats/stats-download-csv/index.jsx#L51

Also I feel like the logic to update the label belongs to `prepareDataForDownload` more.

File: client/my-sites/stats/stats-module-utm/utm-export-button.jsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+			.join( '\n' );
+
+		return csvData;
+	};
+
+	const initiateDownload = ( event ) => {
+		event.preventDefault();
+
+		// TODO: Provide a better default file name.
+		const fileName = 'my-data.csv';
```
Comment (kangzj) [inline client/my-sites/stats/stats-module-utm/utm-export-button.jsx:48]: We could copy and tweak this to start with:

https://github.com/Automattic/wp-calypso/blob/ae401a84d6e9fe28cefc9dc4fd3045f4891d1746/client/my-sites/stats/stats-download-csv/index.jsx#L33

**Review by kangzj — APPROVED** [review]
Works as described! Nice work 👍 

Left two comments but no blockers :shipit:

File: client/my-sites/stats/stats-module-utm/utm-export-button.jsx (hunk 1 above)
Comment (a8ck3n) [reply client/my-sites/stats/stats-module-utm/utm-export-button.jsx:18]: I had a similar feeling but in the `prepareDataForDownload`, we don't have the parent for context anymore. I could try pushing the parent label when flattening the array though. I'll give that, and the double quotes check, a look.

File: client/my-sites/stats/stats-module-utm/utm-export-button.jsx (hunk 2 above)
Comment (a8ck3n) [reply client/my-sites/stats/stats-module-utm/utm-export-button.jsx:48]: That looks like a good starting point! Thanks.

File: client/my-sites/stats/stats-module-utm/utm-export-button.jsx (hunk 1 above)
Comment (a8ck3n) [reply client/my-sites/stats/stats-module-utm/utm-export-button.jsx:18]: Both bits of feedback have been addressed in #88708. 👍

File: client/my-sites/stats/stats-module-utm/utm-export-button.jsx (hunk 2 above)
Comment (a8ck3n) [reply client/my-sites/stats/stats-module-utm/utm-export-button.jsx:48]: Addressed in #88708. Thanks again, Jasper!

**Pull request merged**